kernel/kernel-5.10/drivers/gpio/gpio-thinkedge-se70.c

```diff
@@ -0,0 +1,292 @@
+{
+	int reg, val;
+	struct _se70 *se70_gpio = gpiochip_get_data(gc);
+
+	if (offset >= 4)
+		return;
+
+	mutex_lock(&se70_gpio->lock);
+
+	reg = nct5635y_read(gc, REG_OUTPUT_GPIO1);
+	if (reg < 0)
+		goto out;
+
+	if (value > 0)
+		val = (reg | ser_regs[offset]);
+	else
+		val = (reg & ~ser_regs[offset]);
+
+	nct5635y_write(gc, REG_OUTPUT_GPIO1, val);
+out:
+	mutex_unlock(&se70_gpio->lock);
+}
+
+static int ser_get(struct gpio_chip *gc, unsigned int offset)
+{
+	int reg;
+
+	if (offset >= 4)
+		return -EINVAL;
+
+	reg = nct5635y_read(gc, REG_INPUT_GPIO1);
+	if (reg < 0)
+		return reg;
+
+	if (!(reg & ser_regs[offset]))
+		return 0;
+	else
+		return 1;
+}
+
+static int thinkedge_se70_probe(struct i2c_client *client)
+{
+	struct _se70 *se70_gpio;
+	int ret = 0;
+
+	se70_gpio = devm_kzalloc(&client->dev, sizeof(*se70_gpio), GFP_KERNEL);
+	if (!se70_gpio)
+		return -ENOMEM;
+
+	mutex_init(&se70_gpio->lock);
+
+	/* Digital Input/Output pins */
+	se70_gpio->gpio_dio.label         = "dio";
+	se70_gpio->gpio_dio.parent        = &client->dev;
+	se70_gpio->gpio_dio.owner         = THIS_MODULE;
+	se70_gpio->gpio_dio.get_direction = dio_get_direction;
+	se70_gpio->gpio_dio.get           = dio_get;
+	se70_gpio->gpio_dio.set           = dio_set;
+	se70_gpio->gpio_dio.base          = -1;
+	se70_gpio->gpio_dio.ngpio         = 8;
+	se70_gpio->gpio_dio.can_sleep     = true;
+
+	ret = devm_gpiochip_add_data(&client->dev, &se70_gpio->gpio_dio, se70_gpio);
+	if (ret)
+		goto err_io;
+
+	/* GPIO pins that control Serial Port protocol */
+	se70_gpio->gpio_ser.label         = "serialmode";
+	se70_gpio->gpio_ser.parent        = &client->dev;
+	se70_gpio->gpio_ser.owner         = THIS_MODULE;
+	se70_gpio->gpio_ser.get_direction = ser_get_direction;
+	se70_gpio->gpio_ser.get           = ser_get;
+	se70_gpio->gpio_ser.set           = ser_set;
+	se70_gpio->gpio_ser.base          = -1;
+	se70_gpio->gpio_ser.ngpio         = 4;
+	se70_gpio->gpio_ser.can_sleep     = true;
+
+	ret = devm_gpiochip_add_data(&client->dev, &se70_gpio->gpio_ser, se70_gpio);
+	if (ret)
+		goto err_ser;
+
+	/* Setup Digital I/O pins */
+	ret = nct5635y_write(&se70_gpio->gpio_dio, REG_CONFIG_GPIO0, DIO_CONFIG);
+	if (ret)
+		goto err_setup;
+
+	/* Setup serial port mode pins */
+	ret = setup_serial_pins(&se70_gpio->gpio_ser);
+	if (ret)
+		goto err_setup;
+
+	pr_info("ThinkEdge SE70 GPIO expander initialized\n");
+	return ret;
+err_setup:
+	gpiochip_remove(&se70_gpio->gpio_ser);
+err_ser:
+	gpiochip_remove(&se70_gpio->gpio_dio);
+err_io:
+	pr_err("Failed to initialize ThinkEdge SE70 GPIO expander\n");
+	kfree(se70_gpio);
+	return ret;
+}
+
+static const struct of_device_id i2c_se70_of_match[] = {
+	{ .compatible = "lenovo,thinkedge-se70" },
+	{},
+};
+MODULE_DEVICE_TABLE(of, i2c_se70_of_match);
+
+static struct i2c_driver i2c_se70_driver = {
+	.driver = {
+		.name = "ThinkEdge-SE70",
+		.of_match_table = of_match_ptr(i2c_se70_of_match),
+	},
+	.probe_new = thinkedge_se70_probe,
```

Comment:
Do not see any remove callbacks. Probably missed?

Reply:
it was on purpose actually, I saw a lot of gpio drivers in the kernel that also don't provide remove callbacks and in the case of ThinkEdge SE70, these pins are controlling the serial protocol, shall we really allow the driver removal? I don't have a strong opinion, maybe would be good to allow driver removal... I will think about it....

Reply:
If we build a module, than on module unload I suppose we get a leak, right?

Reply:
no leak, rmmod refuses to unload the module, "resource busy"...

Reply:
I guess this is the magic knowledge that was missing ? so the same register but different bits
`/* I/O pins 0~3 are Digital Outputs and 4~7 are Digital Inputs */`

Reply:
@rouming , I checked the kernel GPIO API and there is no need of remove callback in this case because I'm using the devm_gpiochip_add_data() function, which does the house keeping when the module is removed... the resource busy message I've mentioned occurred because I had the pins exported, so the kernel will not allow the driver removal, the pins must be "unexported" first... so the driver is working and can be removed without problems...